File: Aplicaciones Distribuidas/PARCIAL3/PATRONES DE DISENO VIDEO 3/APLICATIVO/Dotnet/TuberiaYFiltrosGrupo05/TuberiaYFiltrosGrupo05/pipeline/Pipeline.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using TuberiaYFiltrosGrupo05.filters;

namespace TuberiaYFiltrosGrupo05.pipeline
{
    public class Pipeline<I, O>
    {
        private readonly IHandler<I, O> currentHandler;
        private Func<object, object> value;

        public Pipeline(IHandler<I, O> currentHandler)
        {
            this.currentHandler = currentHandler;
        }

        public Pipeline<I, K> AddHandler<K>(IHandler<O, K> newHandler)
        {
            return new Pipeline<I, K>(input => newHandler.Process(this.currentHandler.Process(input)));
        }

        public O Execute(I input)
        {
            return currentHandler.Process(input);
        }
    }

}

File: Aplicaciones Distribuidas/PARCIAL3/PATRONES DE DISENO VIDEO 3/APLICATIVO/Java/TuberiaYFiltrosGrupo05/src/ec/edu/monster/tuberiayfiltrosgrupo05/pipeline/Pipeline.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package ec.edu.monster.tuberiayfiltrosgrupo05.pipeline;

import ec.edu.monster.tuberiayfiltrosgrupo05.filters.Handler;

/**
 *
 * @author santi
 */
public class Pipeline<I, O> {

    private final Handler<I, O> currentHandler;

    public Pipeline(Handler<I, O> currentHandler) {
        this.currentHandler = currentHandler;
    }

    public <K> Pipeline<I, K> addHandler(Handler<O, K> newHandler) {
        return new Pipeline<>(input -> newHandler.process(currentHandler.process(input)));
    }

    public O execute(I input) {
        return currentHandler.process(input);
    }
}